src/rgw/rgw_flight.cc

```diff
@@ -0,0 +1,735 @@
+	return buffer;
+      } else if (file.rdstate() & std::ifstream::failbit &&
+		 file.rdstate() & std::ifstream::eofbit) {
+	const auto bytes_read = file.gcount();
+	INFO << "3 read bytes " << bytes_read << " and reached EOF" << dendl;
+	// buffer->set_size(bytes_read);
+	position = file.tellg();
+	return buffer;
+      } else {
+	ERROR << "unable to read from file" << dendl;
+	return arw::Status::IOError("unable to read from offset %ld", position);
+      }
+    }
+
+    bool supports_zero_copy() const override {
+      return false;
+    }
+
+    // implement Seekable
+
+    arw::Result<int64_t> GetSize() override {
+      return flight_data.obj_size;
+    }
+
+    arw::Result<arw::util::string_view> Peek(int64_t nbytes) override {
+      std::iostream::pos_type here = file.tellg();
+      if (here == -1) {
+	return arw::Status::IOError(
+	  "unable to determine current position ahead of peek");
+      }
+
+      ARROW_ASSIGN_OR_RAISE(OwningStringView result,
+			    OwningStringView::make(nbytes));
+
+      // read
+      ARROW_ASSIGN_OR_RAISE(int64_t bytes_read,
+			    Read(nbytes, (void*) result.writeable_data()));
+      (void) bytes_read; // silence unused variable warnings
+
+      // return offset to original
+      ARROW_RETURN_NOT_OK(Seek(here));
+
+      return result;
+    }
+
+    arw::Result<std::shared_ptr<const arw::KeyValueMetadata>> ReadMetadata() {
+      return flight_data.kv_metadata;
+    }
+
+    arw::Future<std::shared_ptr<const arw::KeyValueMetadata>> ReadMetadataAsync(
+      const arw::io::IOContext& io_context) override {
+      return arw::Future<std::shared_ptr<const arw::KeyValueMetadata>>::MakeFinished(ReadMetadata());
+    }
+
+    // implement Seekable interface
+
+    arw::Status Seek(int64_t position) {
+      file.seekg(position);
+      if (file.fail()) {
+	return arw::Status::IOError(
+	  "error encountered during seek to %" PRId64, position);
+      } else {
+	return arw::Status::OK();
+      }
+    }
+  }; // class LocalRandomAccessFile
+#endif
+
+  class RandomAccessObject : public arw::io::RandomAccessFile {
+
+    FlightData flight_data;
+    const DoutPrefix dp;
+
+    int64_t position;
+    bool is_closed;
+    std::unique_ptr<rgw::sal::Object::ReadOp> op;
+
+  public:
+
+    RandomAccessObject(const FlightData& _flight_data,
+		       std::unique_ptr<rgw::sal::Object>& obj,
+		       const DoutPrefix _dp) :
+      flight_data(_flight_data),
+      dp(_dp),
+      position(-1),
+      is_closed(false)
+    {
+      op = obj->get_read_op();
+    }
+
+    arw::Status Open() {
+      int ret = op->prepare(null_yield, &dp);
+      if (ret < 0) {
+	return arw::Status::IOError(
+	  "unable to prepare object with error %d", ret);
+      }
+      INFO << "file opened successfully" << dendl;
+      position = 0;
+      return arw::Status::OK();
+    }
+
+    // implement InputStream
+
+    arw::Status Close() override {
+      position = -1;
+      is_closed = true;
+      (void) op.release();
```

Comment:
i think you want `reset()` here instead. `release()` transfers ownership to the caller as a raw pointer for the caller to delete themselves, so this would just leak that memory